### PR TITLE
feature: Allow forcing clean checkout of plugins

### PR DESCRIPTION
**Pull request by toothbrush**
## Description

While developing a Buildkite plugin, it can be useful to just work on a branch (e.g., `dev`) and refer to it in some test pipeline, e.g.:

```yaml
steps:
  - command: test.sh
    plugins:
      - my-org/herd-the-llamas#dev:
          alpacas: mostly
```

However, [the current implementation of `bootstrap.go(checkoutPlugin)`](https://github.com/buildkite/agent/blob/de27049657ac876ba08c3e7830c5425d9bb0659d/bootstrap/bootstrap.go#L878) only checks that a checkout matching repo name and ref _exists_, not whether it's up-to-date.

This PR introduces a CLI flag `--plugins-always-clone-fresh` and an environment option `BUILDKITE_PLUGINS_ALWAYS_CLONE_FRESH=true` (thanks @moskyb for the improved names 🙌) which will cause the Buildkite agent to always make a fresh clone of all plugins.

For example, in a test pipeline:

```yaml
steps:
  - command: test.sh
    env:
      BUILDKITE_PLUGINS_ALWAYS_CLONE_FRESH: "true"
    plugins:
      - my-org/herd-the-llamas#dev:
          alpacas: rarely
```

The `BUILDKITE_PLUGINS_ALWAYS_CLONE_FRESH=true` setting will only apply to that one step, which is likely enough for the use case i have in mind.  Of course there are risks that a user of Buildkite might set this option globally, causing their build agents to repeatedly perform clones of plugins on every job step, but.. surely nobody would do that 😅 

Folks have been talking about this for quite a while: https://github.com/buildkite/agent/issues/422!  I'm not sure if this PR adequately addresses the concerns in that issue (e.g., be smart and pull if the plugin version is a branch ref, but not if it's a tag ref), but i've found this to be a sufficient workaround for my purposes.

#### Add a configuration flag per-plugin

Initially i thought it'd be swell to do something like this:

```yaml
steps:
  - command: test.sh
    plugins:
      - my-org/herd-the-llamas#dev:
          force_pull: true
```

However, that configuration block is passed directly to the plugin, and we wouldn't want to block a plugin author from having a `force_pull` flag.  🚫 

#### Add a step-level config flag

Next, i though the "right" place to configure this might be on a step level:

```yaml
steps:
  - command: test.sh
    force_pull_plugins: true
    plugins:
      - my-org/herd-the-llamas#dev:
          alpacas: are_not_llamas
```

But to my surprise, the `buildkite-agent` tool doesn't have much machinery for interpreting pipeline YAML – i guess that happens on Buildkite-server-side.  Fair enough. 🚫 

Then it hit me – the idiomatic way to do this is probably via an environment variable.  I went digging, and discovered inspiration in https://github.com/buildkite/agent/pull/1242.  Crucially, it showed that setting the environment config per-pipeline-step was possible (as opposed to e.g., having to set it before `buildkite-agent` startup, which would be super awkward).  That's why i chose this approach.

### Other random weird comments

* I did quite a bit of thinking and experimenting around making my force-pull feature smarter, e.g., if a checkout exists, do a `git fetch --force --tags` and try to update worktree state to the latest version of the branch or tag.  However, making that robust and general turned out to be surprisingly tricky.  In the end i just chose the blunt instrument and made it `rm -rf` the checkout if the user wants a fresh one.  There's a discussion to be had around that tradeoff, but this seemed most robust and improved the situation today, even if it's not the most ideologically pure way of doing things.
* I was bitten by a very surprising thing (and i might have misunderstood, feel free to correct me): initially i only added the configuration option with environment struct-tag to [`bootstrap/config.go(Config)`](https://github.com/buildkite/agent/blob/de27049657ac876ba08c3e7830c5425d9bb0659d/bootstrap/config.go#L19), but the option only became available to me (as an env var!) once i'd added it with a `cli:"<foo>"` struct tag in [`clicommand/bootstrap.go(BootstrapConfig)`](https://github.com/buildkite/agent/blob/fa992ef4988b6d24820fd17b171c1fceb60ae59c/clicommand/bootstrap.go#L45).  To help future travellers i added a signpost comment in https://github.com/buildkite/agent/pull/1636/commits/5ad83dd8a17daf9f90e9d59e09e60d47ef9a3ce3.
* I wasn't sure how to test this change apart from literally running a build locally with `buildkite-agent bootstrap` and when i had more certainty, spinning up a Buildkite EC2 box with my tricked-out `buildkite-agent`.  So, i haven't got integration tests (yet?) around this feature.  I imagine that `bootstrap/integration/plugin_integration_test.go` would be the place to add something, but i haven't thought about that much yet.
* If folks think this approach has merit, i'd be happy to update the docs with the flag as well as the plugin-development section to suggest using it.

### CODE REVIEW

**Comment by toothbrush** [comment]
Jarryd on the Buildkite slack suggested i ask here.  I've tried to create an integration test for this patch, [here it is](https://github.com/toothbrush/agent/compare/paul/force_pull_plugins...force-pull-integration-test).

It's not acceptable as-is though, because i've struggled to write an integration test that tries to do two test step runs (e.g., run a step, modify the plugin, run the step again and ensure the modified plugin was pulled).  My test runs the Bootstrap bit twice, but (apart from the ugliness of my WIP code) it seems there are mocks or something checking that certain things happen exactly _N_ times, that are failing as a result of the two runs (since perhaps they assume one bootstrap test run per test case).  Do folks have thoughts on how to improve that integration test? (apart from just cleaning it up a bit which i would be happy to do)

**Review by moskyb — CHANGES_REQUESTED** [review]
kia ora @toothbrush!

First off: Thank you so much for the thorough documentation on a) what you've done and b) how you've got there, it's made reviewing this PR a genuine pleasure.

Secondly: This looks really good! i have one minor note about naming, but that's about it in terms of actual substantive changes for this PR. Awesome work.

re: Integration testing: yeah.... i feel your pain. the tool we use to mock external binaries, `bintest`, is really brittle and hugely difficult to debug when it goes wrong. There's a known (to us at least) issue with one of the integration tests (`TestPreExitHooksFireAfterCancel`) fails on macOS boxes for... reasons. I've looked into what's actually going on to fail the test and come back none the wiser, as it works all good in CI.

All this semi-unrelated waffle is to say: Adding integration tests is awesome, but it's also annoying and difficult. If you can't add an integration test within a reasonable amount of time, at that point, it's on us (Buildkite) as the maintainers to make that process easier. If you're banging your head against the wall to add an integration test and you can't get it to work, don't bother for now. We'll fix up the integration testing process.... at some point. Maybe.

File: bootstrap/bootstrap.go
Comment (moskyb) [reply bootstrap/bootstrap.go:935]: yeah, leave it as is for now i reckon. if people are using this and say that it's too slow then we can put some effort into it, but until then i think this is totally fine.

File: bootstrap/config.go
Comment (toothbrush) [reply bootstrap/config.go:96]: Yes, i agree that naming the variable as descriptively & accurately as possible is a good idea. 👍  My suggestion is to rename to:

```golang
	PluginsAlwaysCloneFresh bool `env:"BUILDKITE_PLUGINS_ALWAYS_CLONE_FRESH"`
```

File: bootstrap/bootstrap.go
Comment (toothbrush) [reply bootstrap/bootstrap.go]: Sounds fair 👍

File: bootstrap/config.go
Comment (moskyb) [reply bootstrap/config.go]: sounds good to me!

**Comment by toothbrush** [comment]
> kia ora @toothbrush!

ka mihi @moskyb 😊 (hope i'm saying that right)

> First off: Thank you so much for the thorough documentation on a) what you've done and b) how you've got there, it's made reviewing this PR a genuine pleasure.
> 
> Secondly: This looks really good! i have one minor note about naming, but that's about it in terms of actual substantive changes for this PR. Awesome work.

Thank you for the kind words.  And i totally agree on the naming – i've made a change there.

> re: Integration testing: yeah.... i feel your pain. the tool we use to mock external binaries, `bintest`, is really brittle and hugely difficult to debug when it goes wrong. There's a known (to us at least) issue with one of the integration tests (`TestPreExitHooksFireAfterCancel`) fails on macOS boxes for... reasons. I've looked into what's actually going on to fail the test and come back none the wiser, as it works all good in CI.

It turns out that was exactly what tripped me up, so i've cleaned up my existing integration tests and pushed them to this branch as well.  Let me know what you think.

> All this semi-unrelated waffle is to say: Adding integration tests is awesome, but it's also annoying and difficult. If you can't add an integration test within a reasonable amount of time, at that point, it's on us (Buildkite) as the maintainers to make that process easier. If you're banging your head against the wall to add an integration test and you can't get it to work, don't bother for now. We'll fix up the integration testing process.... at some point. Maybe.

Once i got my head around it it wasn't too bad.  I'm still not 100% satisfied with my own integration tests, notably the change i've made by introducing the `modifyTestPlugin` helper function.  It could be deduped a bit with the `createTestPlugin` function, but that wouldn't be trivial, and since this is "only" test-helper code, i thought i'd avoid gold-plating it.  I am happy to either accept feedback to improve it or leave out this new set of integration tests entirely – up to you.

Thanks for the kind words and good feedback. 🙌

**Review by moskyb — APPROVED** [review]
Awesome work! I'll merge this after we release Agent 3.36, the release is getting a bit chunky as is.

**Comment by toothbrush** [comment]
I'm tempted to say this PR sufficiently addresses the issue originally raised in https://github.com/buildkite/agent/issues/422 that it can be closed, but my approach is more of a dumb workaround than the smarter approach @lox asked for:

> It would be great to fetch the latest version on each job if the plugin isn't a ref or a symbolic ref

What do folks think?  Close up that dangling issue, or leave it as not-yet-completely-solved?  I considered cc'ing all the folks in that thread who had opinions, but i'll avoid spamming folks just yet until i hear further.

**Comment by toothbrush** [comment]
Out of curiosity @moskyb, when roughly do you expect this to be?  The lack of this feature is making my plugin dev a hassle.

**Comment by moskyb** [comment]
i'll merge this now :)

it'll likely be released next time we get critical mass for a release, will likely be less than a month, but could be more.

**Pull request merged**
